Lesson19/nurbasak/script.js

```diff
@@ -0,0 +1,158 @@
+*/
+
+/*
+-----------------------------------------------------------
+  STEP 2: Add Items to the Cart
+-----------------------------------------------------------
+1. Create an `addItem` method in the `ShoppingCart` class.
+2. The method should:
+   - Accept `name`, `price`, and `quantity` as parameters.
+   - Check if the item already exists in the cart.
+     - If it exists, increase the quantity.
+     - Otherwise, add the new item to the `#items` array.
+*/
+
+/*
+-----------------------------------------------------------
+  STEP 3: Remove Items from the Cart
+-----------------------------------------------------------
+1. Add a `removeItem` method to the `ShoppingCart` class.
+2. The method should:
+   - Accept the `name` of the item to remove.
+   - Remove the item from the `#items` array if it exists.
+*/
+
+
+
+/*
+-----------------------------------------------------------
+  STEP 4: Calculate the Total Cost
+-----------------------------------------------------------
+1. Add a `getTotal` method to the `ShoppingCart` class.
+2. The method should:
+   - Calculate and return the total cost of all items in 
+     the cart.
+*/
+
+
+/*
+-----------------------------------------------------------
+  STEP 5: Apply a Discount
+-----------------------------------------------------------
+1. Add an `applyDiscount` method to the `ShoppingCart` class.
+2. The method should:
+   - Accept a discount code (e.g., 'SAVE10', 'SAVE20').
+   - Apply a percentage discount to the total cost if the 
+     code is valid.
+3. Use an object to store discount codes and their values.
+*/
+
+class ShoppingCart {
+  #items;
+
+  constructor() {
+    this.#items = [];
+  }
+
+  viewCart() {
+    console.log('Viewing cart....');
+    for (const item of this.#items) {
+      console.log(item);
+    }
+  }
+
+  addItem(name, price, quantity) {
+    console.log('Adding new item to the cart...');
+    for (const item of this.#items) {
+      if (name === item.name) {
+        console.log('increase the quantity.');
+        item.quantity++;
+        return;
+      }
+    }
+
+    const id = new Date().getMilliseconds();
+    this.#items.push({ id, name, price, quantity });
+  }
+
+  // In real life, we usually use id to find elements or do any manipulations
+  removeItem(name) {
+    console.log('Removing an item from the cart...');
+    for (let i = 0; i < this.#items.length; i++) {
+      const currentElement = this.#items[i];
+      if (name.toLowerCase() === currentElement.name.toLowerCase()) {
+        this.#items.splice(i, 1);
+        return;
+      }
+    }
+  }
+  getTotal() {
+  console.log('Calculating total cost of items in the cart...');
+  let total=0; 
+  for (let item of this.#items) {
+    total += item.price.amount * item.quantity;
+  
+  }
+  console.log(`Total cost is ${total}`);
+  return total;
+}
+
+applyDiscount(code) {
+  console.log('Applying discount code...');
+  const discounts = {
```

Comment:
In real life it would be better to add discounts as a private property on the class itself. This way it's much easier to manage the values (e.g. you don't need to manually update the code hardcoded in a function but instead you could input the values from an Admin console and update them on the class).